### PR TITLE
Add example of optimizing hyperparameters for handwriting recognition

### PR DESCRIPTION
**Pull request by cbkerr**
Thank you @shihkual for this addition for our poster at MIDAS!

### CODE REVIEW

File: projects/flow.pytorch.HyperparameterOptimize/init.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        root="./source/data", train=False, download=True, transform=transform
+    )
+    return
+
+
+def generate_workspace(pr, hyper_params):
+    for sp in cartesian(**hyper_params):
+        job = pr.open_job(sp)
+        if job not in pr:
+            job.init()
```
Comment (b-butler) [inline projects/flow.pytorch.HyperparameterOptimize/init.py:43]:
```suggestion
        pr.open_job(sp).init()
```
Check is unnecessary.

File: projects/flow.pytorch.HyperparameterOptimize/source/workflow.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+):
+    directives = {"nranks": n_gpu, "ngpu": n_gpu, "walltime": walltime}
+    return directives
+
+
+def cpu_directives(
+    walltime: float = 0.5, n_cpu: int = 1, job_workspace_parent: str = "None"
+):
+    directives = {"nranks": n_cpu, "walltime": walltime}
+    return directives
```
Comment (b-butler) [inline projects/flow.pytorch.HyperparameterOptimize/source/workflow.py:19]:
```suggestion
def cpu_directives(walltime: float = 0.5, n_cpu: int = 1):
    return {"nranks": n_cpu, "walltime": walltime}
```

File: projects/flow.pytorch.HyperparameterOptimize/source/labels.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def check_eval_complete(job):
+    if job.doc.get("evaluation_done", None) is None:
+        return False
+    else:
+        if job.doc["evaluation_done"]:
+            return True
+        else:
+            return False
```
Comment (b-butler) [inline projects/flow.pytorch.HyperparameterOptimize/source/labels.py:37]:
```suggestion
    return job.doc.get("evaluation_done", False)
```

File: projects/flow.pytorch.HyperparameterOptimize/source/labels.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def check_train_complete(job):
+    if job.doc.get("train_done", None) is None:
+        return False
+    else:
+        if job.doc["train_done"]:
+            return True
+        else:
+            return False
```
Comment (b-butler) [inline projects/flow.pytorch.HyperparameterOptimize/source/labels.py:27]:
```suggestion
    return job.doc.get("train_done", False)
```

Comment (b-butler) [reply projects/flow.pytorch.HyperparameterOptimize/source/labels.py:27]: Why aren't you using `Project.{pre,post}.true(key)`?

File: projects/flow.pytorch.HyperparameterOptimize/source/labels.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@Project.label
+def status_label(job):
+    check_list = ["train", "evaluation"]
+    check_results = ""
+    for check_point in check_list:
+        if job.doc.get(f"{check_point}_done") is not None:
+            check_results += f"{check_point}_completed, "
+        return check_results
```
Comment (b-butler) [inline projects/flow.pytorch.HyperparameterOptimize/source/labels.py:17]: Faulty logic the last line should be deindented.
```suggestion
    return ", ".join([
        f"{check_point}_completed" for check_point in ("train", "evaluation")
        if job.doc.get(check_point + "_done", False)
    ])
```

File: projects/flow.pytorch.HyperparameterOptimize/source/vae.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    sm = plt.cm.ScalarMappable(cmap=colormap, norm=norm)
+    sm.set_array([])
+
+    plt.figure(figsize=(12, 10))
+    plt.scatter(z1_arr, z2_arr, c=colormap(c_arr), s=5)
+    plt.colorbar(sm, ticks=np.linspace(0, 10, 10, endpoint=False), label="Ground truth")
+    plt.xlabel(r"$z_1$", fontsize=25)
+    plt.ylabel(r"$z_2$", fontsize=25)
+    plt.title(title_name, fontsize=35)
+    plt.savefig(job.fn("latent.jpg"))
```
Comment (b-butler) [inline projects/flow.pytorch.HyperparameterOptimize/source/vae.py:272]: Use object oriented interface.

File: projects/flow.pytorch.HyperparameterOptimize/project.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+PR = signac.get_project()
+TRAIN_WALLTIME = 1
+EVAL_WALLTIME = 0.5
+
+
+@workflow.training_group
+@Project.operation
+# .with_directives(
+#     workflow.cpu_directives(walltime=TRAIN_WALLTIME))
```
Comment (b-butler) [inline projects/flow.pytorch.HyperparameterOptimize/project.py:15]:
```suggestion
@Project.operation(directives=cpu_directives(walltime=TRAIN_WALLTIME))
```

File: projects/flow.pytorch.HyperparameterOptimize/project.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    train_loader, val_loader = vae.load_data(job)
+    vae.fit(job=job, train_loader=train_loader, val_loader=val_loader, device=device)
+    return
+
+
+@workflow.training_group
+@Project.operation
+# with_directives(
+#     workflow.cpu_directives(walltime=EVAL_WALLTIME))
```
Comment (b-butler) [inline projects/flow.pytorch.HyperparameterOptimize/project.py:33]:
```suggestion
@Project.operation(directives=cpu_directives(walltime=EVAL_WALLTIME))
```

**Review by b-butler — APPROVED** [review]

File: projects/flow.pytorch.HyperparameterOptimize/source/vae.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
```
Comment (tommy-waltmann) [inline projects/flow.pytorch.HyperparameterOptimize/source/vae.py:6]: Usually we don't import modules as single capital letters. How about importing it as `functional` or something more descriptive?

**Review by tommy-waltmann — CHANGES_REQUESTED** [review]
Looks to be in good shape, there appear to be CI issues and there is one small thing that I thought I should mention.

File: projects/flow.pytorch.HyperparameterOptimize/source/vae.py (hunk 9 above)
Comment (b-butler) [reply projects/flow.pytorch.HyperparameterOptimize/source/vae.py:6]: I didn't write it and assumed it was a standard like `import numpy as np`, but I can change it.

Comment (tommy-waltmann) [reply projects/flow.pytorch.HyperparameterOptimize/source/vae.py:6]: `nn` is already imported above it, so you could remove the import and use `nn.functional` instead of `F`

**Review by tommy-waltmann — APPROVED** [review]

**Comment by b-butler** [comment]
PyTorch is not yet available and won't until https://github.com/pytorch/pytorch/pull/112882 is merged, so we will have to wait to merge this.
